src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+
+use crate::TransformArgs;
+use expr::{AggregateExpr, AggregateFunc, MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer key is not a subset of inner key
```

Comment:
I'd express this differently to cover more cases. We can always fuse two Reduce operators as long as the outer one doesn't group by the resulting of any aggregation performed by the inner one. For example, IINM this test would prevent the reductions in the example below from being fused, but we can certainly fuse them, and produce the second plan below:

```
materialize=> explain select a / 20 from (select a / 10 as a from a group by a / 10) a group by a / 20;
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.a (u21)+
 | Distinct group=((#0 / 10))    +
 | Distinct group=((#0 / 20))    +
 
(1 row)

materialize=> explain select a / 10 / 20 from a group by a / 10 / 20;
           Optimized Plan            
-------------------------------------
 %0 =                               +
 | Get materialize.public.a (u21)   +
 | Distinct group=(((#0 / 10) / 20))+
 
(1 row)

materialize=> 
```

Given that, I'd check here whether the outer grouping key uses any column which index is greater than the length of the inner grouping key, which means it depends on the result of any aggregation performed by the inner reduction.

Reply:
Thanks! I was planning to follow-up with that. Done now. 😊